### PR TITLE
mgr/dashboard: Administration > Configuration > rgw_dns_name and rgw_dns_s3website_name not updatable via dashboard

### PR DESCRIPTION
**Pull request by nmunet**
Fixes: https://tracker.ceph.com/issues/68976

Signed-off-by: Naman Munet <namanmunet@li-ff83bccc-26af-11b2-a85c-a4b04bfb1003.ibm.com>

**Previously**: users are not able to edit the configuration which have the 'can_update_at_runtime' flag as false.
**Now**: 
1) Users will be able to edit the configuration related with 'rgw' by forcefully editing the configuration if the 'can_update_at_runtime' flag is false for it.
2) After navigating to Administration >> Configuration, by default configuration with filter level:basic is seen

**Fixes Includes**: 
1) by-passing 'can_update_at_runtime' flag for 'rgw' related configurations as the same can be updated at runtime via CLI. Also implemented a warning popup for user to make force edit to rgw related configurations.
2) when navigated to Administration >> Configuration, modified configuration will be seen as we see in cli "ceph config dump", instead of configuration with filter level:basic 

https://github.com/user-attachments/assets/03f06ca1-eae6-4e18-bab0-91763cd73477

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by nizamial09 — CHANGES_REQUESTED** [review]
@nmunet left a few comments here. overall good but there are more room for improvements. thanks

File: src/pybind/mgr/dashboard/controllers/cluster_configuration.py [hunk 1]
```diff
@@ -80,9 +80,9 @@ def filter(self, names=None):
...
         return config_options
 
-    def create(self, name, value):
+    def create(self, name, value, force_edit=False):
         # Check if config option is updateable at runtime
-        self._updateable_at_runtime([name])
+        self._updateable_at_runtime([name], force_edit)
 
         # Update config option
         avail_sections = ['global', 'mon', 'mgr', 'osd', 'mds', 'client']
```
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/controllers/cluster_configuration.py]: removed `avail_sections`

File: src/pybind/mgr/dashboard/controllers/cluster_configuration.py [hunk 2]
```diff
@@ -80,9 +80,9 @@ def filter(self, names=None):
 
         return config_options
 
-    def create(self, name, value):
+    def create(self, name, value, force_edit=False):
```
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/controllers/cluster_configuration.py]: done, thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.html [hunk 3]
```diff
@@ -24,3 +24,13 @@
...
+
+<ng-template #forcePromoteConfirmation>
+  <cd-alert-panel type="warning">Editing this configuration might require restarting the client</cd-alert-panel>
+  <div class="cds-spacing-6"
+       i18n>
+    <strong>
+       Do you want to force the operation?
+    </strong>
+  </div>
+</ng-template>
```
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.html]: followed same approach, thanks

File: src/pybind/mgr/dashboard/controllers/cluster_configuration.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/controllers/cluster_configuration.py]: https://tracker.ceph.com/issues/69068

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.html [hunk 4]
```diff
@@ -24,3 +24,4 @@
     </span>
   </span>
 </ng-template>
+
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.html:27]: unintentional?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.scss [hunk 5]
```diff
@@ -14,3 +16,7 @@
 ::ng-deep cd-configuration td[cdstabledata].wrap {
   word-break: break-all;
 }
+
+.cds-spacing-6 {
+  margin: spacing.$spacing-06;
+}
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.scss:22]: unused?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts [hunk 6]
```diff
@@ -18,6 +18,7 @@ import { AuthStorageService } from '~/app/shared/services/auth-storage.service';
   styleUrls: ['./configuration.component.scss']
 })
 export class ConfigurationComponent extends ListWithDetails implements OnInit {
+  forcePromoteConfirmation: TemplateRef<any>;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts:21]: unused?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts [hunk 7]
```diff
@@ -91,6 +92,7 @@ export class ConfigurationComponent extends ListWithDetails implements OnInit {
   ) {
     super();
     this.permission = this.authStorageService.getPermissions().configOpt;
+    this.permission = this.authStorageService.getPermissions().configOpt;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts:95]: duplicate.
```suggestion
```

**Review by nizamial09 — APPROVED** [review]
Looks good to me, thank you @nmunet

**Comment by nizamial09** [comment]
jenkins test make check

**Comment by nizamial09** [comment]
jenkins test api

**Comment by nizamial09** [comment]
jenkins test submodules

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test windows

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test api

**Review by nizamial09 — APPROVED** [review]

File: qa/tasks/mgr/dashboard/test_rbd.py [hunk 8]
```diff
@@ -869,7 +869,19 @@ def _get_config_by_name(conf_name):
         self.assertEqual(clone_format_version, 2)
         self.assertStatus(200)
 
+        # will take no effect
```
Comment (afreen23) [inline qa/tasks/mgr/dashboard/test_rbd.py:872]: nit: this comment is at least unclear to me if I want to follow. Can we improve ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts [hunk 9]
```diff
@@ -143,7 +143,9 @@ export class ConfigurationComponent extends ListWithDetails implements OnInit {
     if (selection.selected.length !== 1) {
       return false;
     }
-
-    return selection.selected[0].can_update_at_runtime;
+    if ((this.selection.selected[0].name as string).includes('rgw')) {
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts:146]: better to make 'rgw' as constant and reuse

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts [hunk 10]
```diff
@@ -118,7 +121,8 @@ export class ConfigurationFormComponent extends CdForm implements OnInit {
         this.configForm.get('values').get(value.section).setValue(sectionValue);
       });
     }
-
+    this.forceUpdate =
+      !this.response.can_update_at_runtime && response.name.includes('rgw') ? true : false;
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts:125]:
```suggestion
    this.forceUpdate =
      !this.response.can_update_at_runtime && response.name.includes('rgw');
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts [hunk 11]
```diff
@@ -143,12 +147,28 @@ export class ConfigurationFormComponent extends CdForm implements OnInit {
...
 
     return null;
   }
 
+  openCriticalConfirmModal() {
+    this.modalService.show(CriticalConfirmationModalComponent, {
+      buttonText: $localize`Force Edit`,
+      actionDescription: $localize`force edit`,
+      itemDescription: $localize`configuration`,
+      infoMessage: 'Updating this configuration might require restarting the client',
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts:164]: i18n

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts [hunk 12]
```diff
@@ -25,11 +25,27 @@ export class ConfigurationComponent extends ListWithDetails implements OnInit {
   columns: CdTableColumn[];
   selection = new CdTableSelection();
   filters: CdTableColumn[] = [
+    {
+      name: $localize`Modified`,
+      prop: 'modified',
+      filterOptions: ['yes', 'no'],
+      filterInitValue: 'yes',
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts:32]: i18n if possible.

**Review by afreen23 — CHANGES_REQUESTED** [review]
Please fix the signed-off commit, thanks.

This is a good improvement, thanks 👍🏿

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts (hunk 11 above)
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts:164]: its covered from the popup component, so no need here

**Review by afreen23 — APPROVED** [review]

**Comment by nmunet** [comment]
jenkins test make check

**Review by afreen23 — CHANGES_REQUESTED** [review]

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts [hunk 13]
```diff
@@ -13,6 +13,10 @@ import { CdForm } from '~/app/shared/forms/cd-form';
 import { CdFormGroup } from '~/app/shared/forms/cd-form-group';
 import { NotificationService } from '~/app/shared/services/notification.service';
 import { ConfigFormCreateRequestModel } from './configuration-form-create-request.model';
+import { CriticalConfirmationModalComponent } from '~/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component';
+import { ModalCdsService } from '~/app/shared/services/modal-cds.service';
+
+const RGW = 'rgw';
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts:19]: We dont need to define same constant in both files.
We can move it to a generic file and call from that to both places.

This is for normal housekeeping (not blocking PR)

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts [hunk 14]
```diff
@@ -118,7 +123,7 @@ export class ConfigurationFormComponent extends CdForm implements OnInit {
         this.configForm.get('values').get(value.section).setValue(sectionValue);
       });
     }
-
+    this.forceUpdate = !this.response.can_update_at_runtime && response.name.includes(RGW);
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts:126]: Perhaps add a generic helper function and move this alongwith the constant in a generic file calling at both places.
```suggestion
    this.forceUpdate = !this.response.can_update_at_runtime && hasRGW(response.name);
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts [hunk 15]
```diff
@@ -143,7 +145,9 @@ export class ConfigurationComponent extends ListWithDetails implements OnInit {
     if (selection.selected.length !== 1) {
       return false;
     }
-
-    return selection.selected[0].can_update_at_runtime;
+    if ((this.selection.selected[0].name as string).includes(RGW)) {
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration.component.ts:148]:
```suggestion
    if (hasRGW(this.selection.selected[0].name)) {
```

**Review by afreen23 — COMMENTED** [review]
Non blocking comments but good for maintenance.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts (hunk 13 above)
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/configuration/configuration-form/configuration-form.component.ts:19]: will work on these changes in my next PR related to configuration, as we have more work to do here!
Thanks

**Review by afreen23 — APPROVED** [review]

**Pull request merged**